### PR TITLE
Develop: build and install directories + configure script

### PR DESCRIPTION
**Pull request by hrabbel**
- simplified build and install using configure script
- build directory can be chosen by argument to configure script: -DBUILDDIR
- install directory can be chosen by argument to configure script or cmake: -DINSTALLDIR_LEMONADE 
- libraries are build at /path/to/build/directory/lib/  and installed into /path/to/install/directory/lib/LeMonADE 
- binaries are build at /path/to/build/directory/bin/  and installed into /path/to/install/directory/bin/
- headers are installed into /path/to/install/directory/include/LeMonADE/

### CODE REVIEW

File: configure [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+all: ./${BUILDDIR}/Makefile
+	@ \$(MAKE) -C ${BUILDDIR}
+
+./${BUILDDIR}/Makefile:
+	@ (cd ${BUILDDIR} >/dev/null 2>&1 && cmake ${INSTALLDIR_ARGUMENT} ${PWD_PATH})
+
+clean:
+	@- (cd ${BUILDDIR} >/dev/null 2>&1 && cmake .. >/dev/null 2>&1)
+	@- \$(MAKE) --silent -C ${BUILDDIR} clean || true
+	@- \$(RM) ./${BUILDDIR}/Makefile
```
Comment (Bondoki) [inline configure]: Remove line 71-75

File: configure [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+RM    := rm -r
+
+all: ./${BUILDDIR}/Makefile
+	@ \$(MAKE) -C ${BUILDDIR}
+
+./${BUILDDIR}/Makefile:
+	@ (cd ${BUILDDIR} >/dev/null 2>&1 && cmake ${INSTALLDIR_ARGUMENT} ${PWD_PATH})
+
+clean:
+	@- (cd ${BUILDDIR} >/dev/null 2>&1 && cmake .. >/dev/null 2>&1)
```
Comment (Bondoki) [inline configure]: replace: @- (cd ${BUILDDIR} >/dev/null 2>&1 && cmake .. >/dev/null 2>&1)
to: @- (cd ${BUILDDIR} >/dev/null 2>&1)

Comment (Bondoki) [inline configure]: or remove

File: configure [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	@- \$(MAKE) --silent -C ${BUILDDIR} clean || true
+	@- \$(RM) ./${BUILDDIR}/Makefile
+	@- \$(RM) ./${BUILDDIR}/src
+	@- \$(RM) ./${BUILDDIR}/projects
+	@- \$(RM) ./${BUILDDIR}/CMake*
+	@- \$(RM) ./${BUILDDIR}/*.cmake
+
```
Comment (Bondoki) [inline configure:85]: add: .PHONY: clean

File: configure [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+install:
+	@ \$(MAKE) -C ${BUILDDIR} install 
+
```
Comment (Bondoki) [inline configure]: add: .PHONY: install

File: projects/SimpleSimulator/CMakeLists.txt [hunk 5]
```diff
@@ -1,8 +1,10 @@
...
 
 include_directories ("${LEMONADE_DIR}/include/")
-link_directories ("${LEMONADE_DIR}/lib/")
+link_directories ("${CMAKE_BINARY_DIR}/lib/")
 
 add_executable(SimpleSimulator main.cpp)
 
 target_link_libraries(SimpleSimulator LeMonADE)
+
+install(TARGETS SimpleSimulator DESTINATION bin)
```
Comment (Bondoki) [inline projects/SimpleSimulator/CMakeLists.txt]: remove this line!

File: configure [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+BUILDDIR="build"
+
+#parse command line arguments
+for arg in "$@"; do
+    case $arg in
+        -h | --help)
+            usage         
+            exit
+            ;;
+
```
Comment (Bondoki) [inline configure:28]: also provide option -DLEMONADE_TESTS=ON/OFF

File: src/CMakeLists.txt [hunk 7]
```diff
@@ -37,6 +37,6 @@ IF (project_build_static)
 	ADD_LIBRARY(libLeMonADE-static STATIC ${project_SRCS})
 	#TARGET_LINK_LIBRARIES(staticlib ${project_LIBS})
 	SET_TARGET_PROPERTIES(libLeMonADE-static PROPERTIES VERSION "${APPLICATION_VERSION_MAJOR}.${APPLICATION_VERSION_MINOR}" OUTPUT_NAME ${project_BIN} CLEAN_DIRECT_OUTPUT 1)
-	INSTALL(TARGETS libLeMonADE-static DESTINATION lib)
+	INSTALL(TARGETS libLeMonADE-static DESTINATION lib/LeMonADE)
```
Comment (Bondoki) [inline src/CMakeLists.txt]: instead of "lib/LeMonADE" only "lib"

File: projects/Examples/example4/CMakeLists.txt [hunk 8]
```diff
@@ -2,7 +2,7 @@ cmake_minimum_required (VERSION 2.6)
```
Comment (Bondoki) [inline projects/Examples/example4/CMakeLists.txt:3]: replace by:
if (NOT DEFINED LEMONADE_INCLUDE_DIR)
    message("LEMONADE_INCLUDE_DIR is not provided. Build might fail. Use -DLEMONADE_INCLUDE_DIR=/path/to/installed/include/header/LeMonADE/ or install to default location")
endif()

if (NOT DEFINED LEMONADE_LIBRARY_DIR)
    message("LEMONADE_LIBRARY_DIR is not provided. Build might fail. Use -DLEMONADE_LIBRARY_DIR=/path/to/installed/lib/LeMonADE/ or install to default location")
endif()

include_directories (${LEMONADE_INCLUDE_DIR})
link_directories (${LEMONADE_LIBRARY_DIR})

File: CMakeLists.txt [hunk 9]
```diff
@@ -48,39 +48,47 @@ SET (CMAKE_VERBOSE_MAKEFILE 0) # Use 1 for debugging, 0 for release
 #
 # Project Output Paths
 #
-#MESSAGE(${PROJECT_SOURCE_DIR})
-#SET (MAINFOLDER ${PROJECT_SOURCE_DIR})
 SET (LEMONADE_DIR ${PROJECT_SOURCE_DIR})
```
Comment (Bondoki) [inline CMakeLists.txt:51]: add:
SET (LEMONADE_INCLUDE_DIR ${LEMONADE_DIR})
SET (LEMONADE_LIBRARY_DIR "${CMAKE_BINARY_DIR}/lib")

**Comment by hrabbel** [comment]
Ok, I added the suggestions in the last commit.

File: README.md [hunk 10]
```diff
@@ -1,39 +1,81 @@
...
+
+* The options -DINSTALLDIR_LEMONADE and -DBUILDDIR for the configure script are 
+  optional. The build directory defaults to ./build, the install directory defaults
+  to /usr/local/ . The same goes for the option -DINSTALLDIR_LEMONADE when invoking 
+  cmake (the second compilation and installation way).
+* When installing using 'make install', the library is installed to
+  /given/installation/path/lib/LeMonADE/, or if no path was specified to
+  /usr/local/lib/LeMonADE/. Similar for the header files, which go to
+  /given/installation/path/include/LeMonADE/ or /usr/local/include/LeMonADE/.
+  The binaries are installed into /given/installation/path/bin/ or /usr/local/bin/.
```
Comment (Bondoki) [inline README.md:46]: Remove this line. We only install the libs and include

File: README.md [hunk 11]
```diff
@@ -1,39 +1,81 @@
...
+* Install doxygen 
+* Just do for documentation, only:
 
+```sh
     # generates the docs
     mkdir build
     cd build
     cmake ..
     make docs
+```
```
Comment (Bondoki) [inline README.md:77]: add this lines too:

``` sh
     # generates the docs
     ./configure -DINSTALLDIR_LEMONADE=/path/to/install/LeMonADE/ -DBUILDDIR=/path/to/build/
     make docs
```

File: configure [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+	@ \$(MAKE) -C ${BUILDDIR}
+.PHONY: all
+
+./${BUILDDIR}/Makefile:
+	@ (cd ${BUILDDIR} >/dev/null 2>&1 && cmake ${INSTALLDIR_ARGUMENT} ${PWD_PATH})
+
+clean:
+	@- \$(MAKE) --silent -C ${BUILDDIR} clean || true
+.PHONY: clean
+
```
Comment (Bondoki) [inline configure:85]: Add this lines for the documentation:

``` sh
     docs: 
          @  \$(MAKE) -C ${BUILDDIR} docs
     .PHONY : docs
```

File: CMakeLists.txt [hunk 13]
```diff
@@ -48,39 +48,49 @@ SET (CMAKE_VERBOSE_MAKEFILE 0) # Use 1 for debugging, 0 for release
...
+    SET(CMAKE_INSTALL_PREFIX "${INSTALLDIR_LEMONADE}")
+else (DEFINED INSTALLDIR_LEMONADE)
+	message("INSTALLDIR_LEMONADE set to default" ${CMAKE_INSTALL_PREFIX})
+endif()
+
+
+INSTALL(DIRECTORY "${LEMONADE_DIR}/include/" DESTINATION  "include")
 
 #
 # Add Documentation Targets
```
Comment (Bondoki) [inline CMakeLists.txt:96]: replace the doxygen stuff by this lines:

``` sh
     SET (DOC_INPUT_FILE_PATH "${LEMONADE_DIR}/docs/")
     SET (DOC_OUTPUT_FILE_PATH "${INSTALLDIR_LEMONADE}/docs/")

    FIND_PACKAGE (Doxygen)
    IF (DOXYGEN_FOUND)
        MESSAGE("Build documentation with: make docs")
        IF (EXISTS ${DOC_INPUT_FILE_PATH})
            MESSAGE("Existing File documentation with doxygen")
            configure_file(${DOC_INPUT_FILE_PATH}doxygen.conf ${DOC_OUTPUT_FILE_PATH}doxygen.conf @ONLY)
            configure_file(${DOC_INPUT_FILE_PATH}mainpage.dox ${DOC_OUTPUT_FILE_PATH}mainpage.dox @ONLY)
            configure_file(${DOC_INPUT_FILE_PATH}figures/ProgramStructure.jpg         ${DOC_OUTPUT_FILE_PATH}figures/ProgramStructure.jpg @ONLY)
            ADD_CUSTOM_TARGET(
            docs
                ${DOXYGEN_EXECUTABLE} ${DOC_INPUT_FILE_PATH}doxygen.conf
            WORKING_DIRECTORY ${DOC_OUTPUT_FILE_PATH}
                COMMENT "Generating doxygen project documentation." VERBATIM
            )
        ELSE (EXISTS ${DOC_INPUT_FILE_PATH})
            ADD_CUSTOM_TARGET(docs COMMENT "Doxyfile not found. Please generate a doxygen configuration file to use this target." VERBATIM)
        ENDIF (EXISTS ${DOC_INPUT_FILE_PATH})
    ELSE (DOXYGEN_FOUND)
        ADD_CUSTOM_TARGET(docs COMMENT "Doxygen not found. Please install doxygen to use this target." VERBATIM)
    ENDIF (DOXYGEN_FOUND)
```

**Comment by Bondoki** [comment]
Also update "INSTALL"

**Comment by Bondoki** [comment]
Also update Doxygen.conf and main page.

**Comment by hrabbel** [comment]
Closing this, because I am restructuring my repo. New pull request to come

**Pull request closed without merging**